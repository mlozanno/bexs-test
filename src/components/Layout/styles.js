import styled from 'styled-components';

export const StyledLayout = styled.section`
	max-width: 1024px;
	margin: 0 auto;
`;
