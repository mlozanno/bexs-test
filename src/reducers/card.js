import {
	SET_CARD_DATA,
	SET_CARD_TYPE,
	FLIP,
	FLIP_BACK,
	TOGGLE_FLIP,
	VALID_CARD,
	INVALID_CARD,
	SET_INSTALLMENTS,
} from './actions';

export const reducer = (state, action) => {
	const { payload } = action;

	switch (action.type) {
		case SET_CARD_DATA:
			return {
				...state,
				...payload,
			};
		case SET_CARD_TYPE:
			return {
				...state,
				cardType: payload.cardType,
			};
		case FLIP:
			return {
				...state,
				flipped: true,
			};
		case FLIP_BACK:
			return {
				...state,
				flipped: false,
			};
		case TOGGLE_FLIP:
			return {
				...state,
				flipped: !state.flipped,
			};
		case VALID_CARD:
			return {
				...state,
				isValid: true,
			};

		case INVALID_CARD:
			return {
				...state,
				isValid: false,
			};
		case SET_INSTALLMENTS:
			return {
				...state,
				installments: payload.installments,
			};
		default:
			throw new Error(`Unexpected action`);
	}
};
