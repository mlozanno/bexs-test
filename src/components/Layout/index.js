import React from 'react';
import { Router } from 'react-router-dom';

import history from '~/services/history';

import Routes from '~/routes';

import Header from '~/components/Header';

import { StyledLayout } from './styles';

const Layout = () => (
	<Router history={history}>
		<StyledLayout data-testid="layout">
			<Header />
			<Routes />
		</StyledLayout>
	</Router>
);

export default Layout;
