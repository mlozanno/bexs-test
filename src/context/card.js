import React, { createContext, useReducer } from 'react';

import PropTypes from 'prop-types';

import { reducer } from '~/reducers/card';
import { INITIAL_STATE } from '~/reducers/actions';

export const CardContext = createContext();

export const CardProvider = ({ children }) => {
	const contextValue = useReducer(reducer, INITIAL_STATE);
	return (
		<CardContext.Provider value={contextValue}>{children}</CardContext.Provider>
	);
};

CardProvider.propTypes = {
	children: PropTypes.node.isRequired,
};
