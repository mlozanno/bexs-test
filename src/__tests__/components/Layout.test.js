import React from 'react';
import { MemoryRouter } from 'react-router';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import { CardProvider } from '~/context/card';

import Layout from '~/components/Layout';

describe(`<Layout />`, () => {
	it(`should be a function`, () => {
		expect(typeof Layout).toBe(`function`);
	});

	it(`should be render component`, () => {
		const { getByTestId } = render(
			<CardProvider>
				<MemoryRouter>
					<Layout />
				</MemoryRouter>
			</CardProvider>
		);

		const layout = getByTestId(`layout`);

		expect(layout).toMatchSnapshot();
	});

	it(`should be able to render Checkout component when click in logo`, () => {
		const { getByTestId, getByAltText, getByText } = render(
			<CardProvider>
				<MemoryRouter>
					<Layout />
				</MemoryRouter>
			</CardProvider>
		);

		fireEvent.click(getByText(/Checkout/i));

		fireEvent.click(getByAltText(`Bexs logo`));

		const shopPage = getByTestId(`checkout-page`);

		expect(shopPage).toBeTruthy();
	});

	it(`should be able to render Shop component when click in Home link`, () => {
		const { getByTestId, getByText } = render(
			<CardProvider>
				<MemoryRouter>
					<Layout />
				</MemoryRouter>
			</CardProvider>
		);

		fireEvent.click(getByText(/Shop/i));

		const shopPage = getByTestId(`shop-page`);

		expect(shopPage).toBeTruthy();
	});

	it(`should be able to render Checkout component when click in Checkout link`, () => {
		const { getByTestId, getByText } = render(
			<CardProvider>
				<MemoryRouter>
					<Layout />
				</MemoryRouter>
			</CardProvider>
		);

		fireEvent.click(getByText(/Checkout/i));

		const checkoutPage = getByTestId(`checkout-page`);

		expect(checkoutPage).toBeTruthy();
	});

	it(`should be able to render Wallet component when click in Carteira link`, () => {
		const { getByTestId, getByText } = render(
			<CardProvider>
				<MemoryRouter>
					<Layout />
				</MemoryRouter>
			</CardProvider>
		);

		fireEvent.click(getByText(/Carteira/i));

		const walletPage = getByTestId(`wallet-page`);

		expect(walletPage).toBeTruthy();
	});
});
