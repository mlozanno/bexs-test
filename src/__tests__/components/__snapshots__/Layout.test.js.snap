// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> should be render component 1`] = `
<section
  class="sc-AxhUy NWLmU"
  data-testid="layout"
>
  <header
    class="sc-AxhCb hbYNwt"
    data-testid="header"
  >
    <a
      href="/"
    >
      <img
        alt="Bexs logo"
        src="logo.png"
        srcset="logo@2x.png 2x"
      />
    </a>
    <button
      class="sc-AxirZ cglUSK"
      type="button"
    >
      <svg
        fill="currentColor"
        height="40"
        size="40"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="40"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
    <nav
      class="sc-AxjAm liThSi"
    >
      <ul
        class="sc-AxiKw cNRoFp"
        data-testid="nav"
      >
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/checkout"
          >
            Checkout
          </a>
        </li>
        <li>
          <a
            href="/carteira"
          >
            Carteira
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <div
    data-testid="shop-page"
  />
</section>
`;
